controlplane/kubeadm/internal/cluster.go

```diff
@@ -110,7 +135,41 @@ func OlderThan(t *metav1.Time) func(machine *clusterv1.Machine) bool {
 	}
 }
 
-// FilterMachines returns a filtered list of machines
+// SelectedForUpgrade returns a MachineFilter function to find all machines that have the
+// controlplanev1.SelectedForUpgradeLabel set.
+func SelectedForUpgrade() func(machine *clusterv1.Machine) bool {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil || machine.Labels == nil {
+			return false
+		}
+		if _, ok := machine.Labels[controlplanev1.SelectedForUpgradeLabel]; ok {
+			return true
+		}
+		return false
+	}
+}
+
+// UnionFilterMachines returns a filtered list of machines that matches the union of the applied filters
+func UnionFilterMachines(machines []*clusterv1.Machine, filters ...func(machine *clusterv1.Machine) bool) []*clusterv1.Machine {
+	if len(filters) == 0 {
+		return machines
+	}
+	filteredMachines := make([]*clusterv1.Machine, 0, len(machines))
+
+Machine:
```

Comment:
I'd really like to avoid the goto/label here. I have a few options for you to do so:

1) Do a few filters using the already defined machine filter function we have. After that, join the resulting lists and dedupe. May be fairly clean if we add a MachineSet struct.

2) Flip the filter and machine loop to be `for filter in filter: for machine in machines`. If you want the behavior of returning all machines when there are no filters, return the incoming machine list if there are no filters and then run through the for loop.

Reply:
I've refactored how the filtering is done quite a bit now.